File: archiso/airootfs/usr/share/churros/preferences/widgets/row.py
import os
import gi

gi.require_version("Gtk", "4.0")

from gi.repository import Gtk


ICON_DIR = os.path.join(
    os.path.dirname(__file__),
    "..",
    "assets",
    "icons"
)


class Row(Gtk.Button):

    def __init__(
        self,
        title,
        subtitle=None,
        icon=None,
        value=None,
        suffix=None,
        callback=None
    ):

        super().__init__()

        self.callback = callback

        self.add_css_class(
            "row"
        )

        self.set_has_frame(
            False
        )

        content = Gtk.Box(
            orientation=Gtk.Orientation.HORIZONTAL,
            spacing=14
        )

        content.set_margin_top(12)
        content.set_margin_bottom(12)
        content.set_margin_start(14)
        content.set_margin_end(14)

        #
        # Icono
        #

        if icon is not None:

            image = Gtk.Image.new_from_file(

                os.path.join(
                    ICON_DIR,
                    icon
                )

            )

            image.set_pixel_size(
                22
            )

            content.append(
                image
            )

        #
        # Labels
        #

        labels = Gtk.Box(
            orientation=Gtk.Orientation.VERTICAL,
            spacing=2
        )

        labels.set_hexpand(
            True
        )

        title_label = Gtk.Label(
            label=title
        )

        title_label.set_xalign(
            0
        )

        title_label.add_css_class(
            "row-title"
        )

        self._title_label = title_label

        labels.append(
            title_label
        )

        if subtitle is not None:

            subtitle_label = Gtk.Label(
                label=subtitle
            )

            subtitle_label.set_xalign(
                0
            )

            subtitle_label.add_css_class(
                "row-subtitle"
            )

            self._subtitle_label = subtitle_label

            labels.append(
                subtitle_label
            )

        content.append(
            labels
        )

        #
        # Valor
        #

        if value is not None:

            value_label = Gtk.Label(
                label=value
            )

            value_label.add_css_class(
                "row-value"
            )

            content.append(
                value_label
            )

        #
        # Widget derecho
        #

        elif suffix is not None:

            content.append(
                suffix
            )

        self.set_child(
            content
        )

        #
        # Callback
        #

        if self.callback is not None:

            self.connect(
                "clicked",
                self._on_clicked
            )

    def _on_clicked(
        self,
        button
    ):

        self.callback(
            self
        )

    def set_subtitle(
        self,
        text
    ):

        if hasattr(
            self,
            "_subtitle_label"
        ) and self._subtitle_label is not None:

            self._subtitle_label.set_label(text)

    def set_title(
        self,
        text
    ):

        if hasattr(
            self,
            "_title_label"
        ) and self._title_label is not None:

            self._title_label.set_label(text)

File: rust/preferences/src/widgets/row.rs
// ==========================================
// Row — fila de configuración (botón con icono, título, subtítulo, valor)
// (equivalente a widgets/row.py)
// ==========================================

use gtk::prelude::*;

pub struct Row {
    pub button: gtk::Button,
    title_label: gtk::Label,
    subtitle_label: Option<gtk::Label>,
}

impl Row {
    pub fn new(
        title: &str,
        subtitle: Option<&str>,
        icon: Option<&str>,
        value: Option<&str>,
        suffix: Option<&gtk::Widget>,
        callback: Option<Box<dyn Fn(&gtk::Button)>>,
    ) -> Self {
        let button = gtk::Button::new();
        button.add_css_class("row");
        button.set_has_frame(false);

        let content = gtk::Box::new(gtk::Orientation::Horizontal, 14);
        content.set_margin_top(12);
        content.set_margin_bottom(12);
        content.set_margin_start(14);
        content.set_margin_end(14);

        // Icono
        if let Some(icon_name) = icon {
            let image = gtk::Image::from_file(crate::assets::icon_path(icon_name));
            image.set_pixel_size(22);
            content.append(&image);
        }

        // Labels
        let labels = gtk::Box::new(gtk::Orientation::Vertical, 2);
        labels.set_hexpand(true);

        let title_label = gtk::Label::builder()
            .label(title)
            .xalign(0.0)
            .build();
        title_label.add_css_class("row-title");
        labels.append(&title_label);

        let mut subtitle_label = None;
        if let Some(sub) = subtitle {
            let label = gtk::Label::builder()
                .label(sub)
                .xalign(0.0)
                .build();
            label.add_css_class("row-subtitle");
            labels.append(&label);
            subtitle_label = Some(label);
        }

        content.append(&labels);

        // Valor o widget derecho
        if let Some(val) = value {
            let value_label = gtk::Label::new(Some(val));
            value_label.add_css_class("row-value");
            content.append(&value_label);
        } else if let Some(suffix_widget) = suffix {
            content.append(suffix_widget);
        }

        button.set_child(Some(&content));

        if let Some(cb) = callback {
            button.connect_clicked(cb);
        }

        Self {
            button,
            title_label,
            subtitle_label,
        }
    }

    pub fn widget(&self) -> &gtk::Button {
        &self.button
    }

    pub fn set_title(&self, text: &str) {
        self.title_label.set_label(text);
    }

    pub fn set_subtitle(&self, text: &str) {
        if let Some(label) = &self.subtitle_label {
            label.set_label(text);
        }
    }
}
